marlowe-runtime-web/src/Language/Marlowe/Runtime/Web/Types.hs

```diff
@@ -742,3 +746,88 @@ instance ToSchema (ApplyInputsTxEnvelope CardanoTxBody) where
           , ("txBody", txEnvelopeSchema)
           ]
       & required .~ [ "contractId", "transactionId", "txBody" ]
+
+data NetworkId
+  = Mainnet
+  | Testnet Word32
+  deriving (Show, Eq, Ord)
+
+instance ToJSON NetworkId where
+  toJSON Mainnet = String "mainnet"
+  toJSON (Testnet n) = toJSON n
+
+instance FromJSON NetworkId where
+  parseJSON (String "mainnet") = pure Mainnet
+  parseJSON j = Testnet <$> parseJSON j
+
+instance ToSchema NetworkId where
+  declareNamedSchema _ = do
+    let
+      mainnetSchema = mempty
+        & type_ ?~ OpenApiString
+        & enum_ ?~ ["mainnet"]
+    testnetSchema <- declareSchemaRef (Proxy @Word32)
+    pure $ NamedSchema (Just "NetworkId") $ mempty
+      & oneOf ?~ [Inline mainnetSchema, testnetSchema]
+
+data ChainTip
+  = ChainTipGenesis UTCTime
+  | ChainTip BlockHeader UTCTime
+  deriving (Show, Eq, Ord)
+
+data RuntimeStatus = RuntimeStatus
+  { nodeTip :: ChainTip
+  , runtimeChainTip :: ChainTip
+  , runtimeTip :: ChainTip
+  , networkId :: NetworkId
+  , runtimeVersion :: Version
+  } deriving (Show, Eq, Ord)
+
+instance ToJSON ChainTip where
+  toJSON = \case
+    ChainTipGenesis time -> object [ "genesisTimeUTC" .= iso8601Show time ]
+    ChainTip blockHeader time -> object
+      [ "blockHeader" .= blockHeader
+      , "slotTimeUTC" .= iso8601Show time
+      ]
+
+instance FromJSON ChainTip where
+  parseJSON = withObject "ChainTip" \obj -> do
+    genesisTimeUTC <- obj .:? "genesisTimeUTC"
+    blockHeader <- obj .:? "blockHeader"
+    slotTimeUTC <- obj .:? "slotTimeUTC"
+    case (genesisTimeUTC, blockHeader, slotTimeUTC) of
+      (Nothing, Just blockHeader', Just slotTimeUTC') -> pure $ ChainTip blockHeader' slotTimeUTC'
+      (Just genesisTimeUTC', Nothing, Nothing) -> pure $ ChainTipGenesis genesisTimeUTC'
+      _ -> parseFail "Invalid keys, expecting ([\"genesisTimeUTC\"] | [\"blockHeader\", \"slotTimeUTC\"])"
+
+instance FromHttpApiData ChainTip where
+  parseUrlPiece = first T.pack . eitherDecodeStrict . encodeUtf8
+
+instance FromHttpApiData NetworkId where
+  parseUrlPiece = first T.pack . eitherDecodeStrict . encodeUtf8
+
+instance ToParamSchema ChainTip where
+  toParamSchema _ = mempty
+    & oneOf ?~ [Inline genesisSchema, Inline tipSchema]
+    & OpenApi.description ?~ "The latest known point in the chain on a peer."
+    where
+      genesisSchema =  mempty
+        & type_ ?~ OpenApiObject
+        & properties .~
+            [ ("genesisTimeUTC", Inline $ toParamSchema $ Proxy @UTCTime)
+            ]
+        & required .~ [ "genesisTimeUTC" ]
+
+      tipSchema =  mempty
+        & type_ ?~ OpenApiObject
+        & properties .~
+            [ ("blockHeader", Ref $ Reference "BlockHeader")
+            , ("slotTimeUTC", Inline $ toParamSchema $ Proxy @UTCTime)
+            ]
+        & required .~ [ "genesisTimeUTC" ]
+
+instance ToParamSchema NetworkId where
```

Comment:
I like that method of organization, but I'd rather keep things consistent and refactor the existing code in a separate PR instead of introducing random inconsistencies.